File: kinematics_reference/DT/Assets/Scripts/Kinematics/RobotKinematics.cs
using System;
using UnityEngine;

namespace DigitalTwin.Kinematics
{
    public enum SafetyDecision
    {
        Safe,
        Warning,
        Stop
    }

    public readonly struct RobotLinkSegment
    {
        public readonly string Name;
        public readonly int Index;
        public readonly Vector3 Start;
        public readonly Vector3 End;

        public RobotLinkSegment(string name, int index, Vector3 start, Vector3 end)
        {
            Name = string.IsNullOrWhiteSpace(name) ? $"Link{index + 1}" : name;
            Index = index;
            Start = start;
            End = end;
        }
    }

    public readonly struct ForwardKinematicsResult
    {
        public readonly Vector3[] JointPositions;
        public readonly RobotLinkSegment[] LinkSegments;
        public readonly Vector3 EndEffectorPosition;
        public readonly float EndEffectorPlanarAngleDegrees;

        public ForwardKinematicsResult(Vector3[] jointPositions, RobotLinkSegment[] linkSegments, float endEffectorPlanarAngleDegrees)
        {
            JointPositions = jointPositions;
            LinkSegments = linkSegments;
            EndEffectorPosition = jointPositions[jointPositions.Length - 1];
            EndEffectorPlanarAngleDegrees = endEffectorPlanarAngleDegrees;
        }
    }

    public readonly struct HumanRobotDistanceResult
    {
        public readonly bool IsValid;
        public readonly float DistanceMeters;
        public readonly int HumanJointIndex;
        public readonly int RobotLinkIndex;
        public readonly Vector3 HumanJointPosition;
        public readonly Vector3 ClosestPointOnRobotLink;

        public HumanRobotDistanceResult(
            bool isValid,
            float distanceMeters,
            int humanJointIndex,
            int robotLinkIndex,
            Vector3 humanJointPosition,
            Vector3 closestPointOnRobotLink)
        {
            IsValid = isValid;
            DistanceMeters = distanceMeters;
            HumanJointIndex = humanJointIndex;
            RobotLinkIndex = robotLinkIndex;
            HumanJointPosition = humanJointPosition;
            ClosestPointOnRobotLink = closestPointOnRobotLink;
        }
    }

    public readonly struct InverseKinematicsResult
    {
        public readonly bool Converged;
        public readonly float[] JointAnglesDegrees;
        public readonly int Iterations;
        public readonly float ResidualNorm;

        public InverseKinematicsResult(bool converged, float[] jointAnglesDegrees, int iterations, float residualNorm)
        {
            Converged = converged;
            JointAnglesDegrees = jointAnglesDegrees;
            Iterations = iterations;
            ResidualNorm = residualNorm;
        }
    }

    public static class RobotKinematics
    {
        public static ForwardKinematicsResult ForwardKinematics(RobotKinematicParameters parameters, float[] jointAnglesDegrees)
        {
            ValidateParameters(parameters, jointAnglesDegrees);

            Vector3 horizontal = parameters.horizontalAxis.normalized;
            Vector3 vertical = parameters.verticalAxis.normalized;
            Vector3[] joints = new Vector3[parameters.links.Length + 1];
            RobotLinkSegment[] segments = new RobotLinkSegment[parameters.links.Length];
            joints[0] = parameters.basePositionMeters;

            float cumulativeAngleDegrees = 0f;
            for (int i = 0; i < parameters.links.Length; i++)
            {
                RobotLinkParameter link = parameters.links[i];
                cumulativeAngleDegrees += jointAnglesDegrees[i] + link.jointOffsetDegrees;

                // The planar serial-chain model from the Mathematica notebook:
                // p_i = p_{i-1} + L_i [cos(sum theta), sin(sum theta)].
                // The configurable horizontal/vertical axes map that 2D plane into Unity world space.
                float angleRadians = cumulativeAngleDegrees * Mathf.Deg2Rad;
                Vector3 direction = horizontal * Mathf.Cos(angleRadians) + vertical * Mathf.Sin(angleRadians);
                joints[i + 1] = joints[i] + direction * link.linkLengthMeters;
                segments[i] = new RobotLinkSegment(link.name, i, joints[i], joints[i + 1]);
            }

            return new ForwardKinematicsResult(joints, segments, cumulativeAngleDegrees);
        }

        public static HumanRobotDistanceResult MinimumDistance(Vector3[] humanJoints, RobotLinkSegment[] robotSegments)
        {
            if (humanJoints == null || humanJoints.Length == 0 || robotSegments == null || robotSegments.Length == 0)
            {
                return new HumanRobotDistanceResult(false, float.PositiveInfinity, -1, -1, Vector3.zero, Vector3.zero);
            }

            float bestDistance = float.PositiveInfinity;
            int bestHumanJoint = -1;
            int bestRobotLink = -1;
            Vector3 bestHumanPoint = Vector3.zero;
            Vector3 bestRobotPoint = Vector3.zero;

            for (int humanIndex = 0; humanIndex < humanJoints.Length; humanIndex++)
            {
                Vector3 humanPoint = humanJoints[humanIndex];
                if (!IsFinite(humanPoint))
                {
                    continue;
                }

                for (int linkIndex = 0; linkIndex < robotSegments.Length; linkIndex++)
                {
                    float distance = PointToSegmentDistance(humanPoint, robotSegments[linkIndex].Start, robotSegments[linkIndex].End, out Vector3 closest);
                    if (distance < bestDistance)
                    {
                        bestDistance = distance;
                        bestHumanJoint = humanIndex;
                        bestRobotLink = linkIndex;
                        bestHumanPoint = humanPoint;
                        bestRobotPoint = closest;
                    }
                }
            }

            return bestHumanJoint < 0
                ? new HumanRobotDistanceResult(false, float.PositiveInfinity, -1, -1, Vector3.zero, Vector3.zero)
                : new HumanRobotDistanceResult(true, bestDistance, bestHumanJoint, bestRobotLink, bestHumanPoint, bestRobotPoint);
        }

        public static float PointToSegmentDistance(Vector3 point, Vector3 segmentStart, Vector3 segmentEnd, out Vector3 closestPoint)
        {
            Vector3 segment = segmentEnd - segmentStart;
            float lengthSquared = segment.sqrMagnitude;
            if (lengthSquared <= Mathf.Epsilon)
            {
                closestPoint = segmentStart;
                return Vector3.Distance(point, segmentStart);
            }

            // Orthogonal projection parameter clamped to [0, 1] so the closest point lies on the finite link body.
            float t = Vector3.Dot(point - segmentStart, segment) / lengthSquared;
            t = Mathf.Clamp01(t);
            closestPoint = segmentStart + t * segment;
            return Vector3.Distance(point, closestPoint);
        }

        public static SafetyDecision DecideSafety(
            bool dataIsFresh,
            bool gestureStopRequested,
            bool gestureGoRequested,
            float distanceMeters,
            float warningThresholdMeters,
            float stopThresholdMeters,
            float releaseThresholdMeters,
            SafetyDecision previousDecision)
        {
            if (!dataIsFresh || gestureStopRequested)
            {
                return SafetyDecision.Stop;
            }

            if (!IsFinite(distanceMeters))
            {
                return SafetyDecision.Stop;
            }

            if (previousDecision == SafetyDecision.Stop && !gestureGoRequested && distanceMeters < releaseThresholdMeters)
            {
                return SafetyDecision.Stop;
            }

            if (distanceMeters <= stopThresholdMeters)
            {
                return SafetyDecision.Stop;
            }

            return distanceMeters <= warningThresholdMeters ? SafetyDecision.Warning : SafetyDecision.Safe;
        }

        public static InverseKinematicsResult SolvePlanarIk3(
            RobotKinematicParameters parameters,
            Vector2 targetMeters,
            float targetPlanarAngleDegrees,
            float[] initialGuessDegrees,
            int maxIterations = 40,
            float tolerance = 1e-4f)
        {
            if (parameters == null || parameters.links == null || parameters.links.Length != 3)
            {
                throw new ArgumentException("Newton-Raphson IK currently supports exactly the 3-link planar model from the notebook.");
            }

            if (initialGuessDegrees == null || initialGuessDegrees.Length != 3)
            {
                throw new ArgumentException("Initial guess must contain three joint angles.");
            }

            float[] theta = (float[])initialGuessDegrees.Clone();
            float residualNorm = float.PositiveInfinity;

            for (int iteration = 0; iteration < maxIterations; iteration++)
            {
                Vector2 p = PlanarCoordinates(parameters, ForwardKinematics(parameters, theta).EndEffectorPosition);
                float phi = theta[0] + parameters.links[0].jointOffsetDegrees
                    + theta[1] + parameters.links[1].jointOffsetDegrees
                    + theta[2] + parameters.links[2].jointOffsetDegrees;

                float[] residual =
                {
                    p.x - targetMeters.x,
                    p.y - targetMeters.y,
                    Mathf.DeltaAngle(targetPlanarAngleDegrees, phi)
                };

                residualNorm = Mathf.Sqrt(residual[0] * residual[0] + residual[1] * residual[1] + residual[2] * residual[2]);
                if (residualNorm <= tolerance)
                {
                    return new InverseKinematicsResult(true, theta, iteration, residualNorm);
                }

                float[,] jacobian = NumericalJacobian(parameters, theta);
                float[] correction = Solve3x3(jacobian, new[] { -residual[0], -residual[1], -residual[2] });
                for (int i = 0; i < 3; i++)
                {
                    theta[i] += correction[i];
                }
            }

            return new InverseKinematicsResult(false, theta, maxIterations, residualNorm);
        }

        private static Vector2 PlanarCoordinates(RobotKinematicParameters parameters, Vector3 worldPoint)
        {
            Vector3 relative = worldPoint - parameters.basePositionMeters;
            return new Vector2(
                Vector3.Dot(relative, parameters.horizontalAxis.normalized),
                Vector3.Dot(relative, parameters.verticalAxis.normalized));
        }

        private static float[,] NumericalJacobian(RobotKinematicParameters parameters, float[] theta)
        {
            const float h = 1e-3f;
            float[,] jacobian = new float[3, 3];
            Vector2 basePoint = PlanarCoordinates(parameters, ForwardKinematics(parameters, theta).EndEffectorPosition);
            float basePhi = theta[0] + theta[1] + theta[2]
                + parameters.links[0].jointOffsetDegrees
                + parameters.links[1].jointOffsetDegrees
                + parameters.links[2].jointOffsetDegrees;

            for (int column = 0; column < 3; column++)
            {
                float[] perturbed = (float[])theta.Clone();
                perturbed[column] += h;
                ForwardKinematicsResult fk = ForwardKinematics(parameters, perturbed);
                float phi = perturbed[0] + perturbed[1] + perturbed[2]
                    + parameters.links[0].jointOffsetDegrees
                    + parameters.links[1].jointOffsetDegrees
                    + parameters.links[2].jointOffsetDegrees;

                Vector2 planarPoint = PlanarCoordinates(parameters, fk.EndEffectorPosition);
                jacobian[0, column] = (planarPoint.x - basePoint.x) / h;
                jacobian[1, column] = (planarPoint.y - basePoint.y) / h;
                jacobian[2, column] = Mathf.DeltaAngle(basePhi, phi) / h;
            }

            return jacobian;
        }

        private static float[] Solve3x3(float[,] matrix, float[] rhs)
        {
            float[,] a = (float[,])matrix.Clone();
            float[] b = (float[])rhs.Clone();

            for (int pivot = 0; pivot < 3; pivot++)
            {
                int bestRow = pivot;
                float bestValue = Mathf.Abs(a[pivot, pivot]);
                for (int row = pivot + 1; row < 3; row++)
                {
                    float value = Mathf.Abs(a[row, pivot]);
                    if (value > bestValue)
                    {
                        bestValue = value;
                        bestRow = row;
                    }
                }

                if (bestValue <= 1e-8f)
                {
                    throw new InvalidOperationException("Singular IK Jacobian.");
                }

                if (bestRow != pivot)
                {
                    for (int column = pivot; column < 3; column++)
                    {
                        (a[pivot, column], a[bestRow, column]) = (a[bestRow, column], a[pivot, column]);
                    }

                    (b[pivot], b[bestRow]) = (b[bestRow], b[pivot]);
                }

                for (int row = pivot + 1; row < 3; row++)
                {
                    float factor = a[row, pivot] / a[pivot, pivot];
                    for (int column = pivot; column < 3; column++)
                    {
                        a[row, column] -= factor * a[pivot, column];
                    }

                    b[row] -= factor * b[pivot];
                }
            }

            float[] x = new float[3];
            for (int row = 2; row >= 0; row--)
            {
                float sum = b[row];
                for (int column = row + 1; column < 3; column++)
                {
                    sum -= a[row, column] * x[column];
                }

                x[row] = sum / a[row, row];
            }

            return x;
        }

        private static void ValidateParameters(RobotKinematicParameters parameters, float[] jointAnglesDegrees)
        {
            if (parameters == null)
            {
                throw new ArgumentNullException(nameof(parameters));
            }

            if (!parameters.IsValid(out string error))
            {
                throw new ArgumentException(error, nameof(parameters));
            }

            if (jointAnglesDegrees == null || jointAnglesDegrees.Length < parameters.links.Length)
            {
                throw new ArgumentException("Joint angle array must contain one angle per configured link.", nameof(jointAnglesDegrees));
            }
        }

        private static bool IsFinite(Vector3 v)
        {
            return IsFinite(v.x) && IsFinite(v.y) && IsFinite(v.z);
        }

        private static bool IsFinite(float value)
        {
            return !float.IsNaN(value) && !float.IsInfinity(value);
        }
    }
}

File: Kinematics/DT/python/robot_kinematics.py
from __future__ import annotations

from dataclasses import dataclass
import json
import math
from pathlib import Path
from typing import Iterable, Sequence


Vector3 = tuple[float, float, float]


@dataclass(frozen=True)
class RobotLinkParameter:
    name: str
    link_length_meters: float
    joint_offset_degrees: float = 0.0


@dataclass(frozen=True)
class RobotKinematicParameters:
    model_name: str
    base_position_meters: Vector3
    horizontal_axis: Vector3
    vertical_axis: Vector3
    links: tuple[RobotLinkParameter, ...]

    @staticmethod
    def from_json(path: str | Path) -> "RobotKinematicParameters":
        data = json.loads(Path(path).read_text(encoding="utf-8"))
        links = tuple(
            RobotLinkParameter(
                name=item.get("name", f"Link{index + 1}"),
                link_length_meters=float(item["linkLengthMeters"]),
                joint_offset_degrees=float(item.get("jointOffsetDegrees", 0.0)),
            )
            for index, item in enumerate(data["links"])
        )

        return RobotKinematicParameters(
            model_name=data.get("modelName", "unnamed_robot_model"),
            base_position_meters=_vector_from_mapping(data.get("basePositionMeters", {})),
            horizontal_axis=_vector_from_mapping(data.get("horizontalAxis", {"x": 1.0, "y": 0.0, "z": 0.0})),
            vertical_axis=_vector_from_mapping(data.get("verticalAxis", {"x": 0.0, "y": 1.0, "z": 0.0})),
            links=links,
        )


@dataclass(frozen=True)
class RobotLinkSegment:
    name: str
    index: int
    start: Vector3
    end: Vector3


@dataclass(frozen=True)
class ForwardKinematicsResult:
    joint_positions: tuple[Vector3, ...]
    link_segments: tuple[RobotLinkSegment, ...]
    end_effector_position: Vector3
    end_effector_angle_degrees: float


@dataclass(frozen=True)
class HumanRobotDistanceResult:
    valid: bool
    distance_meters: float
    human_joint_index: int
    robot_link_index: int
    human_joint_position: Vector3
    closest_point_on_robot_link: Vector3


@dataclass(frozen=True)
class InverseKinematicsResult:
    converged: bool
    joint_angles_degrees: tuple[float, float, float]
    iterations: int
    residual_norm: float


def forward_kinematics(
    parameters: RobotKinematicParameters,
    joint_angles_degrees: Sequence[float],
) -> ForwardKinematicsResult:
    if len(joint_angles_degrees) < len(parameters.links):
        raise ValueError("joint_angles_degrees must contain one angle per configured link")

    horizontal = _normalize(parameters.horizontal_axis)
    vertical = _normalize(parameters.vertical_axis)
    joints: list[Vector3] = [parameters.base_position_meters]
    segments: list[RobotLinkSegment] = []
    cumulative_angle_degrees = 0.0

    for index, link in enumerate(parameters.links):
        cumulative_angle_degrees += joint_angles_degrees[index] + link.joint_offset_degrees
        angle_radians = math.radians(cumulative_angle_degrees)

        # Same serial planar FK as the Mathematica notebook:
        # p_i = p_{i-1} + L_i [cos(sum theta), sin(sum theta)].
        # horizontal_axis and vertical_axis embed the 2D plane in the project frame.
        direction = _add(
            _scale(horizontal, math.cos(angle_radians)),
            _scale(vertical, math.sin(angle_radians)),
        )
        next_joint = _add(joints[-1], _scale(direction, link.link_length_meters))
        joints.append(next_joint)
        segments.append(RobotLinkSegment(link.name or f"Link{index + 1}", index, joints[-2], next_joint))

    return ForwardKinematicsResult(
        joint_positions=tuple(joints),
        link_segments=tuple(segments),
        end_effector_position=joints[-1],
        end_effector_angle_degrees=cumulative_angle_degrees,
    )


def minimum_human_robot_distance(
    human_joints: Iterable[Vector3],
    robot_segments: Sequence[RobotLinkSegment],
) -> HumanRobotDistanceResult:
    best_distance = math.inf
    best_human_index = -1
    best_link_index = -1
    best_human_point = (0.0, 0.0, 0.0)
    best_robot_point = (0.0, 0.0, 0.0)

    for human_index, human_point in enumerate(human_joints):
        if not _is_finite_vector(human_point):
            continue

        for segment in robot_segments:
            distance, closest = point_to_segment_distance(human_point, segment.start, segment.end)
            if distance < best_distance:
                best_distance = distance
                best_human_index = human_index
                best_link_index = segment.index
                best_human_point = human_point
                best_robot_point = closest

    if best_human_index < 0:
        return HumanRobotDistanceResult(False, math.inf, -1, -1, (0.0, 0.0, 0.0), (0.0, 0.0, 0.0))

    return HumanRobotDistanceResult(
        True,
        best_distance,
        best_human_index,
        best_link_index,
        best_human_point,
        best_robot_point,
    )


def point_to_segment_distance(point: Vector3, segment_start: Vector3, segment_end: Vector3) -> tuple[float, Vector3]:
    segment = _sub(segment_end, segment_start)
    length_squared = _dot(segment, segment)
    if length_squared <= 1e-12:
        return _distance(point, segment_start), segment_start

    # Project onto the infinite line, then clamp to the physical link segment.
    t = _dot(_sub(point, segment_start), segment) / length_squared
    t = max(0.0, min(1.0, t))
    closest = _add(segment_start, _scale(segment, t))
    return _distance(point, closest), closest


def decide_safety(
    data_is_fresh: bool,
    gesture_stop_requested: bool,
    gesture_go_requested: bool,
    distance_meters: float,
    warning_threshold_meters: float,
    stop_threshold_meters: float,
    release_threshold_meters: float,
    previous_decision: str,
) -> str:
    if not data_is_fresh or gesture_stop_requested or not math.isfinite(distance_meters):
        return "Stop"

    if previous_decision == "Stop" and not gesture_go_requested and distance_meters < release_threshold_meters:
        return "Stop"

    if distance_meters <= stop_threshold_meters:
        return "Stop"

    return "Warning" if distance_meters <= warning_threshold_meters else "Safe"


def solve_planar_ik3(
    parameters: RobotKinematicParameters,
    target_xy_meters: tuple[float, float],
    target_angle_degrees: float,
    initial_guess_degrees: Sequence[float],
    max_iterations: int = 40,
    tolerance: float = 1e-4,
) -> InverseKinematicsResult:
    if len(parameters.links) != 3:
        raise ValueError("Newton-Raphson IK supports exactly the 3-link planar model")

    theta = [float(value) for value in initial_guess_degrees[:3]]
    if len(theta) != 3:
        raise ValueError("initial_guess_degrees must contain exactly three angles")

    residual_norm = math.inf
    for iteration in range(max_iterations):
        fk = forward_kinematics(parameters, theta)
        end_effector_xy = _planar_coordinates(parameters, fk.end_effector_position)
        residual = [
            end_effector_xy[0] - target_xy_meters[0],
            end_effector_xy[1] - target_xy_meters[1],
            _delta_angle_degrees(target_angle_degrees, fk.end_effector_angle_degrees),
        ]
        residual_norm = math.sqrt(sum(value * value for value in residual))
        if residual_norm <= tolerance:
            return InverseKinematicsResult(True, tuple(theta), iteration, residual_norm)

        jacobian = _numerical_jacobian(parameters, theta)
        correction = _solve_3x3(jacobian, [-value for value in residual])
        theta = [theta[index] + correction[index] for index in range(3)]

    return InverseKinematicsResult(False, tuple(theta), max_iterations, residual_norm)


def _numerical_jacobian(parameters: RobotKinematicParameters, theta: Sequence[float]) -> list[list[float]]:
    h = 1e-3
    base = forward_kinematics(parameters, theta)
    base_xy = _planar_coordinates(parameters, base.end_effector_position)
    jacobian = [[0.0, 0.0, 0.0] for _ in range(3)]

    for column in range(3):
        perturbed = list(theta)
        perturbed[column] += h
        fk = forward_kinematics(parameters, perturbed)
        xy = _planar_coordinates(parameters, fk.end_effector_position)
        jacobian[0][column] = (xy[0] - base_xy[0]) / h
        jacobian[1][column] = (xy[1] - base_xy[1]) / h
        jacobian[2][column] = _delta_angle_degrees(base.end_effector_angle_degrees, fk.end_effector_angle_degrees) / h

    return jacobian


def _planar_coordinates(parameters: RobotKinematicParameters, world_point: Vector3) -> tuple[float, float]:
    relative = _sub(world_point, parameters.base_position_meters)
    horizontal = _normalize(parameters.horizontal_axis)
    vertical = _normalize(parameters.vertical_axis)
    return (_dot(relative, horizontal), _dot(relative, vertical))


def _solve_3x3(matrix: Sequence[Sequence[float]], rhs: Sequence[float]) -> list[float]:
    a = [list(row) for row in matrix]
    b = list(rhs)

    for pivot in range(3):
        best_row = max(range(pivot, 3), key=lambda row: abs(a[row][pivot]))
        if abs(a[best_row][pivot]) <= 1e-12:
            raise ValueError("singular IK Jacobian")

        if best_row != pivot:
            a[pivot], a[best_row] = a[best_row], a[pivot]
            b[pivot], b[best_row] = b[best_row], b[pivot]

        for row in range(pivot + 1, 3):
            factor = a[row][pivot] / a[pivot][pivot]
            for column in range(pivot, 3):
                a[row][column] -= factor * a[pivot][column]
            b[row] -= factor * b[pivot]

    x = [0.0, 0.0, 0.0]
    for row in range(2, -1, -1):
        numerator = b[row] - sum(a[row][column] * x[column] for column in range(row + 1, 3))
        x[row] = numerator / a[row][row]

    return x


def _vector_from_mapping(data: dict[str, float]) -> Vector3:
    return (float(data.get("x", 0.0)), float(data.get("y", 0.0)), float(data.get("z", 0.0)))


def _normalize(vector: Vector3) -> Vector3:
    magnitude = math.sqrt(_dot(vector, vector))
    if magnitude <= 1e-12:
        raise ValueError("axis vectors must be non-zero")
    return (vector[0] / magnitude, vector[1] / magnitude, vector[2] / magnitude)


def _is_finite_vector(vector: Vector3) -> bool:
    return all(math.isfinite(value) for value in vector)


def _delta_angle_degrees(current: float, target: float) -> float:
    return (target - current + 180.0) % 360.0 - 180.0


def _add(a: Vector3, b: Vector3) -> Vector3:
    return (a[0] + b[0], a[1] + b[1], a[2] + b[2])


def _sub(a: Vector3, b: Vector3) -> Vector3:
    return (a[0] - b[0], a[1] - b[1], a[2] - b[2])


def _scale(a: Vector3, scalar: float) -> Vector3:
    return (a[0] * scalar, a[1] * scalar, a[2] * scalar)


def _dot(a: Vector3, b: Vector3) -> float:
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]


def _distance(a: Vector3, b: Vector3) -> float:
    return math.sqrt(sum((a[index] - b[index]) ** 2 for index in range(3)))
